RFC: Deprecated Attribute for Functions (#48)
luau-lang/rfcs: docs/syntax-attribute-functions-deprecated.md

## Motivation

Roblox already has numerous deprecated Luau APIs. LSP strikes out the names of these APIs in the autocompletion popup in Roblox Studio, their use issues deprecation warnings, and their deprecation status is reflected in Roblox's Creator Hub documentation. Unfortunately, Luau lacks support to leverage this tooling for deprecating custom creator-defined functions. Even for core Luau deprecated functions, `getfenv` and `setfenv`, the linter has hardcoded name comparisons to identify their uses and report deprecation messages. Deprecating new APIs requires ad-hoc modifications to the tooling which can only be performed by the language designers.

This RFC proposes the introduction of a `@deprecated` attribute to mark deprecated Luau APIs. This has multiple benefits:
1. It provides an explicit syntactic marker to identify deprecated functions in Luau code.
2. The LSP interface can display the deprecated API in a visually distinct style in the autocompletion popup.
3. The linter can issue a warning message when the deprecated function is used.

## Design

> The `@deprecated` attribute can take upto two named string parameters. The parameters affect the warning message issued by the linter. The `reason` parameter gives the reason for deprecation and the `use` parameter gives the name of the function that should be used in lieu of the deprecated function. The following table shows the warning message issued by the different styles of this attribute when used on a function named `foo`.

Reviewer: ```suggestion
The `@deprecated` attribute can take up to two named string parameters. The parameters affect the warning message issued by the linter. The `reason` parameter gives the reason for deprecation and the `use` parameter gives the name of the function that should be used in lieu of the deprecated function. The following table shows the warning message issued by the different styles of this attribute when used on a function named `foo`.
```

| Style                                  | Message                                                                |
| -------------------------------------- | -----------------------------------------------------------------------|
| `@deprecated`                          | `"Function 'foo' is deprecated."`                                      |
| `@[deprecated {msg = msg}]`            | `"Function 'foo' is deprecated. <msg>"`                                |
| `@[deprecated {alt = alt}]`            | `"Function 'foo' is deprecated. Use '<alt>' instead."`                 |

> | `@[deprecated {msg = msg, alt = alt}]` | `"Function 'foo' is deprecated. <msg> Use '<alt>' instead.`            |

Reviewer 2: Shouldn't `msg` and `alt` values be strings?

Author: Yes, I meant `msg` and `alt` to be placeholders for this string values. I will replace them with string to avoid confusion.

Reviewer 2: Yeah, most confusing is 'alt' since it could be taken as a symbolic reference of the replacement function.

> The `@deprecated` attribute can only be used on top-level functions. It does not apply recursively to the functions defined within the lexical scope of the attributed function.

Reviewer 2: Do not see the need to restrict to top level.

Author: Conceptually, the analysis feels simpler (tractable?) when you restrict to top-level functions. Also, I would be surprised if inner functions are used as the public API of a module. Having said this, I would not mind generalizing it to functions at any nesting depth.

Reviewer 3: Deprecation status is already a field on `Luau::FunctionType`, so it doesn't actually affect the implementation effort.

This RFC proposes the introduction of a `@deprecated` attribute to mark deprecated Luau APIs. This has multiple benefits:
1. It provides an explicit syntactic marker to identify deprecated functions in Luau code.
2. The LSP interface can display the deprecated API in a visually distinct style in the autocompletion popup.
3. The linter can issue a warning message when the deprecated function is used.

## Design

The `@deprecated` attribute can only be used on named functions and methods. It does not apply recursively to the functions defined within the lexical scope of the attributed function. It can take up to two named string parameters. The parameters affect the warning message issued by the linter. The `reason` parameter gives the reason for deprecation and the `use` parameter gives the name of the function that should be used in lieu of the deprecated function. The following table shows the warning message issued by the different styles of this attribute when used on a function named `foo`.

> | Style                                           | Message                                                                |
> | ----------------------------------------------- | -----------------------------------------------------------------------|
> | `@deprecated`                                   | `"Function 'foo' is deprecated."`                                      |
> | `@[deprecated {reason = string}]`               | `"Function 'foo' is deprecated. <reason>"`                             |
> | `@[deprecated {use = string}]`                  | `"Function 'foo' is deprecated, use '<use>' instead."`                 |
> | `@[deprecated {reason = string, use = string}]` | `"Function 'foo' is deprecated, use '<use>' instead. <reason>"`         |

Reviewer: ```suggestion
| Style                                                 | Message                                                                |
| ----------------------------------------------------- | -----------------------------------------------------------------------|
| `@deprecated`                                         | `"Function 'foo' is deprecated."`                                      |
| `@[deprecated {reason = "my reason"}]`                | `"Function 'foo' is deprecated. <reason>"`                             |
| `@[deprecated {use = "bar()"}]`                       | `"Function 'foo' is deprecated, use '<use>' instead."`                 |
| `@[deprecated {reason = "my reason", use = "bar()"}]` | `"Function 'foo' is deprecated, use '<use>' instead. <reason>"`        |
```

If `foo` is a member of class `bar`, the warning messages above will start as `"Member 'bar.foo' is deprecated ...`.

Consider the following example. Function `printTable` is marked deprecated.

```lua

> @[deprecated {use = "table.foreach(table, print)", 
>               reason = "`printTable` forces tables to be printed in one style; `table.foreach` allows configurable printing."}]

Reviewer: ```suggestion
@[deprecated {
    use = "table.foreach(table, print)", 
    reason = "`printTable` forces tables to be printed in one style; `table.foreach` allows configurable printing.",
}]
```

```

Table and class properties also support deprecated members. In the following example property `m` of table `t` is deprecated.

```lua
local t = {}

@deprecated
function t:m() end
```

## Related Work

Many mainstream languages provide a similar mechanism for deprecating language features. They differ in the kind of program elements that can be deprecated, the ability to convert deprecation warnings to errors, and the ability to specify the version which deprecated the API. 

1. **C++** since C++14 provides a `[[deprecated(msg)]]` attribute that can be applied to declaration of a class, a typedef-name, a variable, a nonstatic data member, a function, a namespace, an enumeration, an enumerator, or a template specialization. The compiler emits a deprecation message, `msg`, when the deprecated program element is used.

2. **C#** provides a `[ObsoleteAttribute(msg, error)]` which can be applied to all program elements except assemblies, modules, parameters, and return values. The `msg` string is emitted by the compiler when the deprecated program element is used in code. The `error` flag converts the warning to an error.

> 3. **Java** provides a `@Deprecated` annotation to deprecate a class, method, or field. The Java compiler issues warnings when these program elements are used. A detailed deprecation message can be supplied by a correspoding javadoc `@deprecated` tag.

Reviewer 2: ```suggestion
3. **Java** provides a `@Deprecated` annotation to deprecate a class, method, or field. The Java compiler issues warnings when these program elements are used. A detailed deprecation message can be supplied by a corresponding javadoc `@deprecated` tag.
```

The description above shows the full-form of the deprecation annotation. In most cases, it is valid to leave the message parameter in favor of a generic deprecation message. Compared to most of these languages, the design proposed for deprecating Luau APIs is simpler. 
1. Deprecation is not supported for non-function program elements because Luau only supports attributes on functions. 
2. There is no way to convert deprecation warnings to errors. 
3. There is no explicit attribute parameter for specifying version string. They can be accommodated in the `reason` parameter. If desired, a future RFC could introduce a `version` parameter to the `@deprecated` attribute.

## Drawbacks

> Adding this attribute increases complexity of code. Once the attribute is released, we cannnot deprecate it.

Reviewer 2: ```suggestion
Adding this attribute increases complexity of code. Once the attribute is released, we cannot deprecate it.
```

3. The linter can issue a warning message when the deprecated function is used.

## Design

The `@deprecated` attribute can take upto two named string parameters. The parameters affect the warning message issued by the linter. The `reason` parameter gives the reason for deprecation and the `use` parameter gives the name of the function that should be used in lieu of the deprecated function. The following table shows the warning message issued by the different styles of this attribute when used on a function named `foo`.

| Style                                           | Message                                                                |
| ----------------------------------------------- | -----------------------------------------------------------------------|
| `@deprecated`                                   | `"Function 'foo' is deprecated."`                                      |
| `@[deprecated {reason = string}]`               | `"Function 'foo' is deprecated. <reason>"`                             |
| `@[deprecated {use = string}]`                  | `"Function 'foo' is deprecated. Use '<use>' instead."`                 |

> | `@[deprecated {reason = string, use = string}]` | `"Function 'foo' is deprecated. Use '<use>' instead. <reason>`         |

Reviewer: ```suggestion
| `@[deprecated {reason = string, use = string}]` | `"Function 'foo' is deprecated. Use '<use>' instead. <reason>"`        |
```

Author: Fixed
